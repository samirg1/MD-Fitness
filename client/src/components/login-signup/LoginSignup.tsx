import ArrowCircleRightIcon from "@mui/icons-material/ArrowCircleRight";
import Alert from "@mui/material/Alert";
import Box from "@mui/material/Box";
import Grid from "@mui/material/Grid";
import IconButton from "@mui/material/IconButton";
import { useEffect, useState } from "react";
import Loader from "../Loader";
import Field, { FieldType } from "./Field";
import LoginSignupButton from "./LoginSignupButton";

const LoginSignup = () => {
    const [loggingIn, setLoggingIn] = useState(true);

    const [name, setName] = useState("");
    const [email, setEmail] = useState("");
    const [password, setPassword] = useState("");
    const [fieldsDisabled, setFieldsDisabled] = useState(false);

    const [loading, setLoading] = useState(false);
    const [loginSignupError, setLoginSignupError] = useState("");

    useEffect(() => setLoginSignupError(""), [name, email, password]);

    const toggleLoggingIn = () => {
        setLoggingIn((previous) => !previous);
        setName("");
        setEmail("");
        setPassword("");
    };

    const submit = () => {
        setFieldsDisabled(true);
        setLoading(true);
    };

    return (
        <>
            <Loader isLoading={loading} />
            <Box
                style={{
                    position: "absolute",
                    top: "15%",
                    left: "10%",
                    right: "10%",
                }}
            >
                <Grid container spacing={3}>
                    <Grid item xs={12}>
                        <LoginSignupButton
                            text="Login"
                            disabled={loggingIn}
                            onClick={toggleLoggingIn}
                        />
                        <LoginSignupButton
                            text="Signup"
                            disabled={!loggingIn}
                            onClick={toggleLoggingIn}
                        />
                    </Grid>
                    {!loggingIn ? (
                        <Grid item xs={12}>
                            <Field
                                name="Enter name:"
                                value={name}
                                setValue={setName}
                                disabled={fieldsDisabled}
                                type={FieldType.text}
                            />
                        </Grid>
                    ) : null}
                    <Grid item xs={12}>
                        <Field
                            name="Enter email:"
                            value={email}
                            setValue={setEmail}
                            disabled={fieldsDisabled}
                            type={FieldType.email}
                        />
                    </Grid>
                    <Grid item xs={12}>
                        <Field
                            name="Enter password:"
                            value={password}
                            setValue={setPassword}
                            disabled={fieldsDisabled}
                            type={FieldType.password}
                        />
                        <IconButton onClick={submit} disabled={loading}>
                            <ArrowCircleRightIcon />
                        </IconButton>
                    </Grid>
                    {loginSignupError ? (
                        <Grid item xs={12}>
                            <Alert severity="error">{loginSignupError}</Alert>
                        </Grid>
                    ) : null}
                </Grid>
            </Box>
        </>
    );
};

export default LoginSignup;
