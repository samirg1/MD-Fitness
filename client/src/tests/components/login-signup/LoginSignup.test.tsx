import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import LoginSignup from "../../../components/login-signup/LoginSignup";

const keyDownReturn: any[] = [() => {}];
const accountReturn: any[] = [
    {
        signup: () => {},
        login: () => {},
    },
];
const snackbarReturn: any[] = [{ setOptions: () => {} }];

jest.mock("../../../hooks/useKeyDownHandler", () => () => keyDownReturn[0]);
jest.mock("../../../hooks/useSnackBar", () => () => snackbarReturn[0]);
jest.mock("../../../hooks/useAccount", () => () => accountReturn[0]);
jest.mock("../../../components/login-signup/AccountVerification", () => () => (
    <div>Verify</div>
));

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useLocation: () => {
        return { state: undefined };
    },
    useNavigate: () => {
        return () => {};
    },
}));

describe("basic renders", () => {
    test("intial render", () => {
        const { asFragment } = render(<LoginSignup />);
        expect(asFragment()).toMatchSnapshot();
    });

    test("check toggle button", () => {
        render(<LoginSignup />);
        expect(screen.queryByText("Enter name:")).toBeFalsy();
        expect(
            (screen.getByText("Sign Up") as HTMLButtonElement).disabled
        ).toEqual(false);
        expect(
            (screen.getByText("Login") as HTMLButtonElement).disabled
        ).toBeTruthy();
        fireEvent.click(screen.getByText("Sign Up"));
        expect(screen.getByText("Enter name:")).toBeTruthy();
        expect(
            (screen.getByText("Sign Up") as HTMLButtonElement).disabled
        ).toBeTruthy();
        expect(
            (screen.getByText("Login") as HTMLButtonElement).disabled
        ).toEqual(false);
    });

    test("error message shows and clears", async () => {
        const mockLogin = jest.fn((obj) => "testError");
        accountReturn[0].login = mockLogin;

        render(<LoginSignup />);
        const submitButton = screen.getAllByRole("button")[3];

        fireEvent.click(submitButton);
        await screen.findByText("testError");

        const input = screen.getByRole("textbox");
        fireEvent.change(input, { target: { value: "a" } });

        await waitFor(() =>
            expect(screen.queryByText("testError")).toBeFalsy()
        );
    });
});
