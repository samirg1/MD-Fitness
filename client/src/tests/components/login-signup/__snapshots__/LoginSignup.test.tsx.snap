// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic renders intial render 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
    style="position: absolute; top: 15%; left: 10%; right: 10%;"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiToggleButtonGroup-root css-1gjgmky-MuiToggleButtonGroup-root"
          role="group"
        >
          <button
            aria-pressed="true"
            class="MuiButtonBase-root Mui-disabled MuiToggleButton-root Mui-selected Mui-disabled MuiToggleButton-sizeMedium MuiToggleButton-primary MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-d9c359-MuiButtonBase-root-MuiToggleButton-root"
            disabled=""
            tabindex="-1"
            type="button"
            value="login"
          >
            Login
          </button>
          <button
            aria-pressed="false"
            class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-primary MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-d9c359-MuiButtonBase-root-MuiToggleButton-root"
            tabindex="0"
            type="button"
            value="signup"
          >
            Sign Up
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          Enter email:
        </p>
        <div
          class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <div
            aria-label="Enter a valid email"
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            data-mui-internal-clone-element="true"
          >
            <input
              aria-invalid="true"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              type="email"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-ihdtdm"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          Enter password:
        </p>
        <div
          class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <div
            aria-label="Minimum 8 characters with a lowercase and uppercase letter, a number and a symbol"
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
            data-mui-internal-clone-element="true"
          >
            <input
              aria-invalid="true"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              type="password"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="VisibilityIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-ihdtdm"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowCircleRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 12c0-5.52-4.48-10-10-10S2 6.48 2 12s4.48 10 10 10 10-4.48 10-10zm-10 1H8v-2h4V8l4 4-4 4v-3z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
